parla/parray/coherence.py

```diff
@@ -98,30 +100,37 @@ def read(self, device_id: int) -> MemoryOperation:
 
         Return:
             MemoryOperation, which is a load operation with src and dst
+
+        Note: lock will be acquired
         """
-        device_local_state = self._local_states[device_id]
-
-        if device_local_state == self.INVALID:  # need to load data from somewhere
-            if self._global_state == self.SHARED:
-                self._local_states[device_id] = self.SHARED
-                # load data from owner
-                return MemoryOperation.load(dst=device_id, src=self._owner)
-            elif self._global_state == self.MODIFIED:
-                prev_owner = self._owner
-                self._local_states[prev_owner] = self.SHARED
-                self._local_states[device_id] = self.SHARED
-                self._global_state = self.SHARED
-
-                # Trick: smaller one becomes owner, so will always load from CPU (-1) when possible
-                self._owner = max(self._owner, device_id)
-
-                return MemoryOperation.load(dst=device_id, src=prev_owner)
-            else:   # overall_state should not be INVALID here
-                return MemoryOperation.error()
-        else:
-            return MemoryOperation.noop()  # do nothing
-
-    def write(self, device_id: int) -> List[MemoryOperation]:
+        with self._lock:
+            device_local_state = self._local_states[device_id]
+
+            if device_local_state == self.INVALID:  # need to load data from somewhere
+                if self._global_state == self.SHARED:
+                    # load data from owner
+                    self._local_states[device_id] = self.SHARED
+
+                    self._data_ready[device_id] = False  # copy in progress
+                    return MemoryOperation.load(dst=device_id, src=self._owner)
+                elif self._global_state == self.MODIFIED:
+                    prev_owner = self._owner
+                    self._local_states[prev_owner] = self.SHARED
+                    self._local_states[device_id] = self.SHARED
+                    self._global_state = self.SHARED
+
+                    # Trick: smaller one becomes owner, so will always load from CPU (-1) when possible
+                    self._owner = max(self._owner, device_id)

```

Comment:
this is max. so comment is wrong.

Reply:
Yep. This is a holdover from the old parray code. I didn't touch this but will update